accepted/2024/tensors/README.md

```diff
@@ -0,0 +1,424 @@
+
+`Tensor<T>` is notably explicitly sealed as extensibility must be an intentional design point. The extensibility point here is then `ITensor<TSelf, T>` and `Tensor<T>` is only our own friendly wrapper over arbitrary memory. This allows it to be more efficient and also more easily control lifetimes, ownership, and more freely extend the type in the future as needed.
+
+```csharp
+namespace System.Numerics.Tensors;
+
+public sealed class Tensor<T>
+    : IDisposable,
+      ITensor<Tensor<T>, T>
+    where T : IEquatable<T>, IEqualityOperators<T, T, bool>
+{
+    // IDisposable
+
+    void Dispose();
+
+    // ITensor
+
+    public static Tensor<T> Empty { get; }
+
+    public bool IsEmpty { get; }
+    public bool IsPinned { get; }
+    public int Rank { get; }
+    public ReadOnlySpan<nint> Strides { get; }
+
+    public ref T this[params ReadOnlySpan<nint> indices] { get; }
+
+    public static implicit operator TensorSpan<T>(Tensor<T> value);
+    public static implicit operator TensorReadOnlySpan<T>(Tensor<T> value);
+
+    public TensorSpan<T> AsSpan(params ReadOnlySpan<Range<nint>> ranges);
+    public ref T GetPinnableReference();
+    public Tensor<T> Slice(params ReadOnlySpan<Range<nint>> ranges);
+
+    public void Clear<T>();
+    public void CopyTo<T>(TensorSpan<T> destination);
+    public void Fill<T>(T value);
+    public bool TryCopyTo<T>(TensorSpan<T> destination);
+
+    // IEnumerable
+
+    public Enumerator GetEnumerator();
+
+    // IEquatable
+
+    public bool Equals(Tensor<T> other);
+
+    // IEqualityOperators
+
+    public static bool operator ==(Tensor<T> left, Tensor<T> right);
+    public static bool operator ==(Tensor<T> left, T right);
+
+    public static bool operator !=(Tensor<T> left, Tensor<T> right);
+    public static bool operator !=(Tensor<T> left, T right);
+
+    public struct Enumerator
+    {
+        public ref readonly T Current { get; }
+
+        public bool MoveNext();
+    }
+}
+
+public static partial class Tensor
+{
+    // Effectively mirror the TensorPrimitives surface area. Following the general pattern
+    // where we will return a new tensor and an overload that takes in the destination explicitly.
+    // * public static ITensor<T> Abs<T>(ITensor<T> x) where T : INumberBase<T>;
+    // * public static void Abs<T>(ReadOnlyTensorSpan<T> x, TensorSpan<T> destination) where T : INumberBase<T>;
+
+    // The language ideally has extension operators such that we can define `operator +` instead of `Add`
+    // Without this support, we end up having to have `TensorNumber<T>`, `TensorBinaryInteger<T>`, etc
+    // as we otherwise cannot correctly expose the operators based on what `T` supports
```

Comment:
Is that really the fallback? I'd have expected the fallback would be named extension methods until extension operators are supported, at which point we would add operators that map to the named methods. It'd be unfortunate if we shipped types we knew were going to be immediate legacy.

Reply:
I didn't explain this well enough. The proposal is meant to go with what you say. Exposing explicit `Add` methods and let that eventually become `extension operator +`. If we don't do that, the alternative is to do what generic math did, which while the correct choice for the fundamental numeric interfaces, is much less ideal for the general purpose patterns built on top of that, like `Tensor<T>`.